File: android/src/main/kotlin/com/haishinkit/haishin_kit/RtmpConnectionHandler.kt
package com.haishinkit.haishin_kit

import android.os.Handler
import android.os.Looper
import com.haishinkit.event.Event
import com.haishinkit.event.IEventListener
import com.haishinkit.rtmp.RtmpConnection
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class RtmpConnectionHandler(
    private val plugin: HaishinKitPlugin,
    private val id: Int,
) : MethodChannel.MethodCallHandler, IEventListener,
    EventChannel.StreamHandler {
    companion object {
        private const val TAG = "RtmpConnection"
    }

    var instance: RtmpConnection? = RtmpConnection()
        private set

    private var channel: EventChannel
    private var eventSink: EventChannel.EventSink? = null

    init {
        instance?.addEventListener(Event.RTMP_STATUS, this)
        channel = EventChannel(
            plugin.flutterPluginBinding.binaryMessenger,
            "com.haishinkit.eventchannel/${id}"
        )
        channel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "$TAG#connect" -> {
                val command = call.argument<String>("command") ?: ""
                instance?.connect(command)
                result.success(null)
            }
            "$TAG#close" -> {
                instance?.close()
                result.success(null)
            }
            "$TAG#dispose" -> {
                eventSink?.endOfStream()
                instance?.dispose()
                instance = null
                plugin.onDispose(id)
                result.success(null)
            }
        }
    }

    override fun handleEvent(event: Event) {
        val map = HashMap<String, Any?>()
        map["type"] = event.type
        map["data"] = event.data
        plugin.uiThreadHandler.post {
            eventSink?.success(map)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
    }
}

File: ios/Classes/RTMPConnectionHandler.swift
import Foundation
import Flutter
import HaishinKit

final class RTMPConnectionHandler: NSObject, MethodCallHandler {
    var instance: RTMPConnection?
    private let plugin: HaishinKitPlugin
    private var channel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var subscription: Task<(), Error>? {
        didSet {
            oldValue?.cancel()
        }
    }

    init(plugin: HaishinKitPlugin) {
        self.plugin = plugin
        super.init()
        let id = Int(bitPattern: ObjectIdentifier(self))
        if let messanger = plugin.registrar?.messenger() {
            self.channel = FlutterEventChannel(name: "com.haishinkit.eventchannel/\(id)", binaryMessenger: messanger)
        } else {
            self.channel = nil
        }
        instance = RTMPConnection()
        channel?.setStreamHandler(self)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        Task {
            switch call.method {
            case "RtmpConnection#connect":
                guard
                    let arguments = call.arguments as? [String: Any?],
                    let command = arguments["command"] as? String else {
                    return
                }
                if let instance {
                    subscription = Task {
                        for await status in await instance.status {
                            eventSink?(status.makeEvent())
                        }
                    }
                }
                _ = try? await instance?.connect(command)
                result(nil)
            case "RtmpConnection#close":
                subscription = nil
                try? await instance?.close()
                result(nil)
            case "RtmpConnection#dispose":
                instance = nil
                plugin.onDispose(id: Int(bitPattern: ObjectIdentifier(self)))
                result(nil)
            default:
                result(FlutterMethodNotImplemented)
            }
        }
    }
}

extension RTMPConnectionHandler: FlutterStreamHandler {
    // MARK: FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}
